File: Dynamic-Programming/123-H-Best-Time-to-Buy-and-Sell-Stock-III/best_time_to_buy_and_sell_stock_iii.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        state1, state2, state3, state4 = -prices[0], float('-inf'), float('-inf'), float('-inf') 
        for price in prices:
            state1 = max(state1, -price)
            state2 = max(state2, state1+price)
            state3 = max(state3, state2-price)
            state4 = max(state4, state3+price)
        return max(state4, 0) 

File: Dynamic-Programming/123-H-Best-Time-to-Buy-and-Sell-Stock-III/best_time_to_buy_and_sell_stock_iii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if (n<=1) return 0;

        int state1 = -prices[0];
        int state2 = INT_MIN;
        int state3 = INT_MIN;
        int state4 = INT_MIN;

        for (int i=0; i<n; i++) {
            state1 = max(state1, -prices[i]);
            state2 = max(state2, state1+prices[i]);
            state3 = max(state3, state2-prices[i]);
            state4 = max(state4, state3+prices[i]);
        }
        return max(state4, 0);
    }
};